clang/lib/Sema/SemaChecking.cpp

```diff
@@ -179,6 +179,39 @@ static bool SemaBuiltinCHERICapCreate(Sema &S, CallExpr *TheCall) {
   return false;
 }
 
+/// Check that argument \p ArgIndex is a capability type (or an array/function
+/// that decays to a capability type.
+static bool checkCapArg(Sema &S, CallExpr *TheCall, unsigned ArgIndex,
+                        QualType *ResultingSrcTy = nullptr) {
+  clang::Expr *Source = TheCall->getArg(ArgIndex);
+  ASTContext &Ctx = S.Context;
+  QualType SrcTy = Source->getType();
+  // We should also be able to use it with arrays/functions
+  if (SrcTy->canDecayToPointerType())
+    SrcTy = Ctx.getDecayedType(SrcTy);
+
+  // Also allow NULL/nullptr in hybrid mode to be promoted to a capability
+  if (SrcTy->isNullPtrType())
+    SrcTy = Ctx.getPointerType(Ctx.VoidTy, ASTContext::PIK_Capability);
+  else if (SrcTy->isPointerType() && !SrcTy->isCapabilityPointerType() &&
```

Comment:
I think you just need to drop the `SrcTy->isPointerType()` from here to permit a literal uncast 0; looking at `isNullPointerConstant` it handles the plain integer 0 at the end.

Reply:
Yeah, I just assumed we wanted to be explicit for a literal zero whether the result should be void* or intcap_t.

Reply:
Well, either you require a cast on the input all the time, or you only require a cast on the input (or output) some of the time, and since `void * __capability` feels like the less surprising default (due to `void *` being more "core" to C than the more-recent `(u)intptr_t`) then we should not require casts in those cases IMO. There's no wholly-satisfactory answer, but I think that's the best we can do.